### ISSUE TITLE
Thread don't execute immediately  until join? 

### ISSUE THREAD

**Issue by cagev** [alibaba/PhotonLibOS#615]
The threads don't execute immediately ,   until I press enter , and invoke the join() method. 

```cpp

#include <thread>
#include <chrono>

#include <photon/thread/thread.h>
#include <photon/thread/std-compat.h>
#include <photon/thread/thread11.h>

int main(int argc, char *argv[])
{
    photon::init();
    std::this_thread::sleep_for(std::chrono::seconds(1));
    printf("...............init.................\n");
    auto thrd = photon_std::thread([=]()
                                   { printf("execute thread 1...............\n"); });
    auto thrd2 = photon::thread_create11([=]()
                                         { printf("execute thread ....2...........\n"); });
    photon_std::thread thrd3([=]()
                             { printf("execute thread 3 .................\n"); });

    thrd3.detach(); 
    printf("...............done........................\n");

    getchar();  //didn't execute the thread task, until I press enter here 

    thrd.join();  //execute from here? 

    
    // thrd2->join();

    return 0;
}

```

execute result : 

08:21:17 arthur@WIPC photonthread ±|main ✗|→ ./bin/testthread
2024/10/30 20:21:18.685225|INFO |th=0000556191962E20|epoll.cpp:318|new_epoll_engine:Init epoll event engine: master
2024/10/30 20:21:18.685250|DEBUG|th=0000556191962E20|reset_handle.cpp:27|ResetHandle:push [this=0000556191960850]
2024/10/30 20:21:18.685272|DEBUG|th=0000556191962E20|reset_handle.cpp:27|ResetHandle:push [this=0000556191960B20]
2024/10/30 20:21:18.685284|INFO |th=0000556191962E20|signal.cpp:292|sync_signal_init:signalfd initialized
2024/10/30 20:21:18.685313|INFO |th=0000556191962E20|epoll.cpp:318|new_epoll_engine:Init epoll event engine: cascading
2024/10/30 20:21:18.685325|DEBUG|th=0000556191962E20|reset_handle.cpp:27|ResetHandle:push [this=000055619195E370]
2024/10/30 20:21:18.685873|DEBUG|th=0000556191962E20|curl.cpp:243|libcurl_init:libcurl version libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
2024/10/30 20:21:18.685903|DEBUG|th=0000556191962E20|reset_handle.cpp:27|ResetHandle:push [this=0000556191964150]
2024/10/30 20:21:18.685904|INFO |th=0000556191962E20|curl.cpp:261|libcurl_init:libcurl initialized
2024/10/30 20:21:18.685921|DEBUG|th=0000556191962E20|reset_handle.cpp:27|ResetHandle:push [this=0000556191964030]
2024/10/30 20:21:18.685922|DEBUG|th=0000556191962E20|aio-wrapper.cpp:398|libaio_wrapper_init:libaio initialized
2024/10/30 20:21:18.685925|DEBUG|th=0000556191962E20|photon.cpp:106|__photon_init:reset_all_handle registed [getpid()=1296389]
...............init.................
...............done........................

//here I press enter 
execute thread 1...............
execute thread ....2...........
execute thread 3 .................

env: 
wsl with gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

tested using the latest version and v0.8.1

**Comment by beef9999** [comment]
> No thread yield in this call, which means the caller will continue its execution.
> https://photonlibos.github.io/docs/api/thread#description

Unlike std::thread, photon's thread runs in vCPU. Your program has only one vCPU, and the scheduler has no chance to make the context switch to another photon thread.

std::threads will create real OS threads, so it behaves differently.

**Comment by lihuiba** [comment]
Photon's threads are cooperatively scheduled (coroutine), similar to virtual thread in Java. It is a must for each thread to explicitly yield (give up the control of vCPU), either directly or indirectly, so that other thread(s) in the same vCPU can get executed. 

The join function internally yields, so the new thread gets executed.
